### PR TITLE
Avoid using Get interface to obtain metadata when iterating data during migration

### LINKED ISSUE #904: Avoid using Get interface to obtain Metadata when iterating data during migration
### Motivation

When migrating data, I found that when migrating the same number of keys (about 10W), the migration time was very variable, sometimes very short, sometimes very long. When the migration took a long time, I found that the CPU was full. 

Looking at the flame diagram below:

image

we can see that there is a **deep call stack** on the left. 

Zoomed in it:

image

When we parsed and sent the full data (`SlotMigrate::SendSnapshot()`), we used the `rocksdb::DB::Get()` interface to get the metadata, which was CPU consuming.

### How much does this affect migration speed?

1. Because we set the iterator to not fill the cache, when the amount of data is larger than cache, it is very likely that we will call `rocksdb::DB::Get()` to read disk, which greatly affects the performance!
2. If the retrieval in memtable is bad( it depends on workload), then it will consume more CPU, which will bring more serious impact.

### Solution

In fact, we do not need to get the value through the `rocksdb::DB::Get()` interface. When we iterate data, value and key are together, so we can directly get the value through the iterator and then encode into metadata.

### PR DESCRIPTION
**Pull request by caipengbo**
Fixed #904

### CODE REVIEW

File: src/slot_migrate.cc [hunk 1]
```diff
@@ -302,7 +302,7 @@ Status SlotMigrate::SendSnapshot(void) {
 
     // Add key's constructed cmd to restore_cmds, send pipeline
     // or not according to current_pipeline_size_
-    auto stat = MigrateOneKey(rocksdb::Slice(user_key), &restore_cmds);
+    auto stat = MigrateOneKey(rocksdb::Slice(user_key), iter->value(), &restore_cmds);
```
Comment (git-hulk) [inline src/slot_migrate.cc]:
```suggestion
    auto stat = MigrateOneKey(user_key, iter->value(), &restore_cmds);
```

**Review by git-hulk — DISMISSED** [review]

**Comment by datavisorxiaobiaozhao** [comment]
The problem we had was that when we migrated, we had a lot of disk reads, but with this optimized version, these abnormal disk reads disappeared, so I think GetMetaData calls the Get interface

**Comment by caipengbo** [comment]
> The problem we had was that when we migrated, we had a lot of disk reads, but with this optimized version, these abnormal disk reads disappeared, so I think GetMetaData calls the Get interface.

@datavisorxiaobiaozhao Yes. During the iteration, we set the cache not to be filled:

https://github.com/apache/incubator-kvrocks/blob/e8721a270d23345f385af6db0617c3549d6e6812/src/slot_migrate.cc#L273-L277

so that when we call `DB::Get()`, we are likely to read the disk.

File: src/slot_migrate.cc [hunk 2]
```diff
@@ -543,29 +543,15 @@ bool SlotMigrate::CheckResponseWithCounts(int sock_fd, int total) {
...
-}
-
-Status SlotMigrate::MigrateOneKey(rocksdb::Slice key, std::string *restore_cmds) {
+Status SlotMigrate::MigrateOneKey(const rocksdb::Slice &key, const rocksdb::Slice &value, std::string *restore_cmds) {
   std::string prefix_key;
   AppendNamespacePrefix(key, &prefix_key);
-  std::string bytes;
-  bool st = GetSlotKeyMetadata(prefix_key, &bytes);
-  if (!st) return Status(Status::NotOK, "[migrate] Failed to get key's metadata");
+  std::string bytes = value.ToString();
```
Comment (xiaobiaozhao) [inline src/slot_migrate.cc:549]: Optimization big value
```suggestion
  std::string bytes = std::move(value.ToString());
```

Comment (PragmaTwice) [reply src/slot_migrate.cc:549]: This `std::move` is useless, and it will block the (N)RVO ([return value optimization](https://en.cppreference.com/w/cpp/language/copy_elision)).

explainations:

###  1. this `std::move` is useless
Signature of `ToString`:
```
std::string Slice::ToString(...);
```
So the return value `x.ToString()` is already an rvalue (or concretely, a [prvalue](https://en.cppreference.com/w/cpp/language/value_category#prvalue)).
Move an rvalue is useless, since it will essentially dispatch to the rvalue reference overload.

### 2. it will block the (N)RVO
RVO and NRVO are important optimization approaches in C++, and are guaranteed in new standards.

So, in `std::string x = v.ToString()`, the local result variable of `ToString` will be eliminated and it just construct one `std::string x`, no copy contruction, and even no move construction.

If you move it, then RVO will just not work here.

So, in conclusion, this change will only downgrade the performance.

Comment (caipengbo) [reply src/slot_migrate.cc:549]: That's great. That's something I don't know about. I'll dig into it. thank you @PragmaTwice

**Review by git-hulk — APPROVED** [review]

File: src/slot_migrate.cc (hunk 2 above)
Comment (xiaobiaozhao) [reply src/slot_migrate.cc:549]: That's great!!!

**Comment by shangxiaoxiong** [comment]
There may be some problem. The Get interface may aim to get the latest expire time of the key.

**Comment by caipengbo** [comment]
> There may be some problem. The Get interface may aim to get the latest expire time of the key.

@shangxiaoxiong Oh Yes, I missed that. Then we only need GetMetaData for those with expiration time. Or we can record the time snapshot was created and subtract the expiration time to determine whether it is expired!

**Comment by shangxiaoxiong** [comment]
During sending snapshot, exipre time of some key may be changed to larger by clients. GetMetaData is needed to deal with this case.

**Comment by shangxiaoxiong** [comment]
l think GetMetaData is needed for complex key with expire time to get the latest expire time.

**Comment by caipengbo** [comment]
> During sending snapshot, exipre time of some key may be changed to larger by clients. GetMetaData is needed to deal with this case.

During sending snapshot:
- for simple type, user changes to expire are incremental data and will be synchronized through WAL.
- for complex types, we really should get the latest metadata.

**Comment by shangxiaoxiong** [comment]
> > During sending snapshot, exipre time of some key may be changed to larger by clients. GetMetaData is needed to deal with this case.
> 
> During sending snapshot, user changes to expire are incremental data and will be synchronized through WAL.

Simple key is ok but complex key is not.

**Comment by git-hulk** [comment]
> > > During sending snapshot, exipre time of some key may be changed to larger by clients. GetMetaData is needed to deal with this case.
> > 
> > 
> > During sending snapshot, user changes to expire are incremental data and will be synchronized through WAL.
> 
> Simple key is ok but complex key is not.

It can't resolve this issue even we use the latest snapshot since the change may come after.

**Comment by shangxiaoxiong** [comment]
Yeah. Using GetMetaData with latest snapshot can't solve the problem completely. Just a little improvement.

**Comment by caipengbo** [comment]
Indeed, the whole process is not atomic.

**Comment by git-hulk** [comment]
Yes, the root cause is the snapshot can be different between writers and background threads. I prefer to use current implementation, then we can try to seek another way to resolve this issue completely. How do you think? @caipengbo @shangxiaoxiong

**Comment by shangxiaoxiong** [comment]
Current implementation makes slotmigration faster.  Faster slotmigration is, smaller risk of losing key is.  On the whole, current implementation may be good.

**Comment by caipengbo** [comment]
I think OK, I will continue to think about how to solve this kind of problem.

**Comment by xiaobiaozhao** [comment]
I think this is the best solution for now, merge this pr first, and then contionue to look for new solutions.

**Review by xiaobiaozhao — APPROVED** [review]

**Review by shangxiaoxiong — APPROVED** [review]

**Review by git-hulk — APPROVED** [review]

**Comment by caipengbo** [comment]
Thanks all for the great discussion, I'll explain the problems with this implementation in #904

**Review by PragmaTwice — APPROVED** [review]
